File: lib/services/lease_contract_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/lease_contract_model.dart';
import '../utils/constants.dart';
import 'storage_service.dart';

class LeaseContractService {
   Future<String?> _getToken() async {
    return await StorageService.getToken();
  }
  Future<Map<String, String>> _getHeaders() async {
    final token = await _getToken();
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  Future<LeaseContractModel> createContract({
    required String apartmentId,
    required String ownerId,
    required String tenantId,
    required DateTime startDate,
    DateTime? endDate,
    required double initialRentAmount,
    double? depositAmount,
    double? chargesAmount,
    required String regionCode,
  }) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('${Constants.baseUrl}/lease-contracts'),
      headers: headers,
      body: jsonEncode({
        'apartmentId': apartmentId,
        'ownerId': ownerId,
        'tenantId': tenantId,
        'startDate': startDate.toIso8601String().split('T')[0],
        'endDate': endDate?.toIso8601String().split('T')[0],
        'initialRentAmount': initialRentAmount,
        'depositAmount': depositAmount,
        'chargesAmount': chargesAmount,
        'regionCode': regionCode,
      }),
    );

    if (response.statusCode == 200) {
      return LeaseContractModel.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to create contract: ${response.body}');
    }
  }

  Future<List<LeaseContractModel>> getMyContracts() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('${Constants.baseUrl}/lease-contracts/my-contracts'),
      headers: headers,
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((json) => LeaseContractModel.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load contracts: ${response.body}');
    }
  }

  Future<LeaseContractModel> getContractById(String contractId) async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('${Constants.baseUrl}/lease-contracts/$contractId'),
      headers: headers,
    );

    if (response.statusCode == 200) {
      return LeaseContractModel.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load contract: ${response.body}');
    }
  }

  Future<LeaseContractModel> signContractByOwner(String contractId, String signatureData) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('${Constants.baseUrl}/lease-contracts/$contractId/sign-owner'),
      headers: headers,
      body: jsonEncode({'signatureData': signatureData}),
    );

    if (response.statusCode == 200) {
      return LeaseContractModel.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to sign contract: ${response.body}');
    }
  }

  Future<LeaseContractModel> signContractByTenant(String contractId, String signatureData) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('${Constants.baseUrl}/lease-contracts/$contractId/sign-tenant'),
      headers: headers,
      body: jsonEncode({'signatureData': signatureData}),
    );

    if (response.statusCode == 200) {
      return LeaseContractModel.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to sign contract: ${response.body}');
    }
  }

  Future<RentIndexationModel> indexRent({
    required String contractId,
    required double indexationRate,
    double? baseIndex,
    double? newIndex,
    String? notes,
  }) async {
    final headers = await _getHeaders();
    final queryParams = {
      'indexationRate': indexationRate.toString(),
      if (baseIndex != null) 'baseIndex': baseIndex.toString(),
      if (newIndex != null) 'newIndex': newIndex.toString(),
      if (notes != null) 'notes': notes,
    };

    final response = await http.post(
      Uri.parse('${Constants.baseUrl}/lease-contracts/$contractId/index-rent').replace(queryParameters: queryParams),
      headers: headers,
    );

    if (response.statusCode == 200) {
      return RentIndexationModel.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to index rent: ${response.body}');
    }
  }

  Future<String> generateContractPdf(String contractId) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('${Constants.baseUrl}/lease-contracts/$contractId/generate-pdf'),
      headers: headers,
    );

    if (response.statusCode == 200) {
      return response.body;
    } else {
      throw Exception('Failed to generate PDF: ${response.body}');
    }
  }
}

File: src/main/java/be/delomid/oneapp/mschat/mschat/service/LeaseContractService.java
package be.delomid.oneapp.mschat.mschat.service;

import be.delomid.oneapp.mschat.mschat.dto.*;
import be.delomid.oneapp.mschat.mschat.model.*;
import be.delomid.oneapp.mschat.mschat.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class LeaseContractService {

    private final LeaseContractRepository leaseContractRepository;
    private final ApartmentRepository apartmentRepository;
    private final ResidentRepository residentRepository;
    private final RentIndexationRepository rentIndexationRepository;
    private final LeaseContractCustomSectionRepository customSectionRepository;
    private final PdfLeaseContractGenerationService pdfGenerationService;

    @Transactional
    public LeaseContractDto createContract(CreateLeaseContractRequest request) {
        Apartment apartment = apartmentRepository.findById(request.getApartmentId())
                .orElseThrow(() -> new RuntimeException("Apartment not found"));

        Resident owner = residentRepository.findById(request.getOwnerId())
                .orElseThrow(() -> new RuntimeException("Owner not found"));

        Resident tenant = residentRepository.findById(request.getTenantId())
                .orElseThrow(() -> new RuntimeException("Tenant not found"));

        if (!apartment.getOwner().getIdUsers().equals(owner.getIdUsers())) {
            throw new RuntimeException("Only the owner of the apartment can create a lease contract");
        }

        if (leaseContractRepository.findByApartment_IdApartmentAndStatus(
                request.getApartmentId(), LeaseContractStatus.SIGNED).isPresent()) {
            throw new RuntimeException("Cannot create a new contract while an active contract exists for this apartment");
        }

        boolean isOwnerOccupant = owner.getIdUsers().equals(tenant.getIdUsers());

        LeaseContract contract = LeaseContract.builder()
                .apartment(apartment)
                .owner(owner)
                .tenant(tenant)
                .startDate(request.getStartDate())
                .endDate(request.getEndDate())
                .initialRentAmount(request.getInitialRentAmount())
                .currentRentAmount(request.getInitialRentAmount())
                .depositAmount(request.getDepositAmount())
                .chargesAmount(request.getChargesAmount())
                .regionCode(request.getRegionCode())
                .status(isOwnerOccupant ? LeaseContractStatus.SIGNED : LeaseContractStatus.DRAFT)
                .build();

        if (isOwnerOccupant) {
            contract.setOwnerSignedAt(LocalDateTime.now());
            contract.setTenantSignedAt(LocalDateTime.now());
        }

        contract = leaseContractRepository.save(contract);

        apartment.setTenant(tenant);
        if (isOwnerOccupant || contract.getStatus() == LeaseContractStatus.SIGNED) {
            apartment.setResident(tenant);
        }
        apartmentRepository.save(apartment);

        return convertToDto(contract);
    }

    public List<LeaseContractDto> getContractsByOwner(String ownerId) {
        return leaseContractRepository.findByOwner_IdUsers(ownerId)
                .stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    public List<LeaseContractDto> getContractsByTenant(String tenantId) {
        return leaseContractRepository.findByTenant_IdUsers(tenantId)
                .stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    public LeaseContractDto getContractById(UUID contractId) {
        LeaseContract contract = leaseContractRepository.findById(contractId)
                .orElseThrow(() -> new RuntimeException("Contract not found"));
        return convertToDto(contract);
    }

    @Transactional
    public LeaseContractDto signContractByOwner(UUID contractId, String signatureData) {
        throw new RuntimeException("Direct contract signature is not allowed. Please complete and sign the entry inventory (état des lieux d'entrée) first. The contract will be automatically signed once the entry inventory is completed.");
    }

    @Transactional
    public LeaseContractDto signContractByTenant(UUID contractId, String signatureData) {
        throw new RuntimeException("Direct contract signature is not allowed. Please complete and sign the entry inventory (état des lieux d'entrée) first. The contract will be automatically signed once the entry inventory is completed.");
    }

    @Transactional
    public RentIndexationDto indexRent(UUID contractId, BigDecimal indexationRate, BigDecimal baseIndex, BigDecimal newIndex, String notes) {
        LeaseContract contract = leaseContractRepository.findById(contractId)
                .orElseThrow(() -> new RuntimeException("Contract not found"));

        BigDecimal previousAmount = contract.getCurrentRentAmount();
        BigDecimal newAmount = previousAmount.multiply(BigDecimal.ONE.add(indexationRate));

        RentIndexation indexation = RentIndexation.builder()
                .contract(contract)
                .indexationDate(LocalDate.now())
                .previousAmount(previousAmount)
                .newAmount(newAmount)
                .indexationRate(indexationRate)
                .baseIndex(baseIndex)
                .newIndex(newIndex)
                .notes(notes)
                .build();

        indexation = rentIndexationRepository.save(indexation);

        contract.setCurrentRentAmount(newAmount);
        leaseContractRepository.save(contract);

        return convertIndexationToDto(indexation);
    }

    public String generateContractPdf(UUID contractId) {
        try {
            String pdfUrl = pdfGenerationService.generateLeaseContractPdf(contractId);

            LeaseContract contract = leaseContractRepository.findById(contractId)
                    .orElseThrow(() -> new RuntimeException("Contract not found"));

            contract.setPdfUrl(pdfUrl);
            leaseContractRepository.save(contract);

            return pdfUrl;
        } catch (Exception e) {
            log.error("Error generating PDF for contract: {}", contractId, e);
            throw new RuntimeException("Failed to generate PDF: " + e.getMessage(), e);
        }
    }

    private LeaseContractDto convertToDto(LeaseContract contract) {
        log.info( contract.getOwner().getFname() + " " + contract.getOwner().getLname() +"//////"+
                  contract.getTenant().getFname() + " " + contract.getTenant().getLname());
        return LeaseContractDto.builder()
                .id(contract.getId().toString())
                .apartmentId(contract.getApartment().getIdApartment())
                .ownerId(contract.getOwner().getIdUsers())
                .ownerName(contract.getOwner().getFname() + " " + contract.getOwner().getLname())
                .tenantName(contract.getTenant().getFname() + " " + contract.getTenant().getLname())
                .tenantId(contract.getTenant().getIdUsers())
                .startDate(contract.getStartDate())
                .endDate(contract.getEndDate())
                .initialRentAmount(contract.getInitialRentAmount())
                .currentRentAmount(contract.getCurrentRentAmount())
                .depositAmount(contract.getDepositAmount())
                .chargesAmount(contract.getChargesAmount())
                .regionCode(contract.getRegionCode())
                .status(contract.getStatus().name())
                .ownerSignedAt(contract.getOwnerSignedAt())
                .tenantSignedAt(contract.getTenantSignedAt())
                .pdfUrl(contract.getPdfUrl())
                .createdAt(contract.getCreatedAt())
                .updatedAt(contract.getUpdatedAt())
                .build();
    }

    private RentIndexationDto convertIndexationToDto(RentIndexation indexation) {
        return RentIndexationDto.builder()
                .id(indexation.getId().toString())
                .contractId(indexation.getContract().getId().toString())
                .indexationDate(indexation.getIndexationDate())
                .previousAmount(indexation.getPreviousAmount())
                .newAmount(indexation.getNewAmount())
                .indexationRate(indexation.getIndexationRate())
                .baseIndex(indexation.getBaseIndex())
                .newIndex(indexation.getNewIndex())
                .notes(indexation.getNotes())
                .createdAt(indexation.getCreatedAt())
                .build();
    }
}
